File: python/app/services/token_service.py
import uuid
from datetime import datetime, timedelta
from hashlib import sha256

from jose import jwt, JWTError

from app.config import settings
from app.database import SessionLocal
from app.models.refresh_token import RefreshToken


def create_access_token(data: dict) -> str:
    to_encode = data.copy()
    expire = datetime.utcnow() + timedelta(minutes=settings.access_token_expiry_minutes)
    to_encode.update({"exp": expire})
    return jwt.encode(to_encode, settings.JWT_ACCESS_SECRET, algorithm="HS256")


def create_refresh_token(user_id: str) -> str:
    raw_token = str(uuid.uuid4())
    hashed_token = sha256(raw_token.encode()).hexdigest()
    expire = datetime.utcnow() + timedelta(minutes=settings.refresh_token_expiry_minutes)

    db = SessionLocal()
    try:
        rt = RefreshToken(
            user_id=user_id,
            token=hashed_token,
            expires_at=expire,
        )
        db.add(rt)
        db.commit()
    finally:
        db.close()

    return raw_token


def verify_refresh_token(token: str) -> RefreshToken | None:
    hashed_token = sha256(token.encode()).hexdigest()

    db = SessionLocal()
    try:
        rt = (
            db.query(RefreshToken)
            .filter(
                RefreshToken.token == hashed_token,
                RefreshToken.revoked == False,
                RefreshToken.expires_at > datetime.utcnow(),
            )
            .first()
        )
        return rt
    finally:
        db.close()


def revoke_refresh_token(token: str) -> None:
    hashed_token = sha256(token.encode()).hexdigest()

    db = SessionLocal()
    try:
        rt = db.query(RefreshToken).filter(RefreshToken.token == hashed_token).first()
        if rt:
            rt.revoked = True
            db.commit()
    finally:
        db.close()


def revoke_all_user_tokens(user_id: str) -> None:
    db = SessionLocal()
    try:
        tokens = db.query(RefreshToken).filter(RefreshToken.user_id == user_id, RefreshToken.revoked == False).all()
        for rt in tokens:
            rt.revoked = True
        db.commit()
    finally:
        db.close()

    return token


def verify_access_token(token: str) -> dict:
    try:
        payload = jwt.decode(token, settings.JWT_ACCESS_SECRET, algorithms=["HS256"])
        return payload
    except JWTError:
        return None


def verify_refresh_token(token: str) -> RefreshToken | None:
    token_hash = sha256(token.encode()).hexdigest()

    db = SessionLocal()
    try:
        rt = (
            db.query(RefreshToken)
            .filter(
                RefreshToken.token_hash == token_hash,
                RefreshToken.revoked == False,
                RefreshToken.expires_at > datetime.utcnow(),
            )
            .first()
        )
        return rt
    finally:
        db.close()


def revoke_refresh_token(token: str) -> None:
    token_hash = sha256(token.encode()).hexdigest()

    db = SessionLocal()
    try:
        rt = db.query(RefreshToken).filter(RefreshToken.token_hash == token_hash).first()
        if rt:
            rt.revoked = True
            db.commit()
    finally:
        db.close()


def revoke_all_user_tokens(user_id: str) -> None:
    db = SessionLocal()
    try:
        tokens = db.query(RefreshToken).filter(RefreshToken.user_id == user_id, RefreshToken.revoked == False).all()
        for rt in tokens:
            rt.revoked = True
        db.commit()
    finally:
        db.close()

File: node/src/services/tokenService.js
const jwt = require('jsonwebtoken');
const crypto = require('crypto');
const { RefreshToken } = require('../models');

function generateAccessToken(user) {
  return jwt.sign(
    { id: user.id, email: user.email, name: user.name },
    process.env.JWT_ACCESS_SECRET,
    { expiresIn: process.env.JWT_ACCESS_EXPIRES_IN || '15m' }
  );
}

async function generateRefreshToken(user) {
  const token = jwt.sign(
    { id: user.id },
    process.env.JWT_REFRESH_SECRET,
    { expiresIn: process.env.JWT_REFRESH_EXPIRES_IN || '7d' }
  );

  const expiresAt = new Date();
  expiresAt.setDate(expiresAt.getDate() + 7);

  const hashedToken = crypto.createHash('sha256').update(token).digest('hex');

  await RefreshToken.create({
    userId: user.id,
    token: hashedToken,
    expiresAt,
  });

  return token;
}

function verifyAccessToken(token) {
  return jwt.verify(token, process.env.JWT_ACCESS_SECRET);
}

async function verifyRefreshToken(token) {
  const payload = jwt.verify(token, process.env.JWT_REFRESH_SECRET);
  const hashedToken = crypto.createHash('sha256').update(token).digest('hex');

  const tokenRecord = await RefreshToken.findOne({
    where: { token: hashedToken, revoked: false },
  });

  if (!tokenRecord) {
    throw new Error('Refresh token not found or revoked');
  }

  if (new Date() > new Date(tokenRecord.expiresAt)) {
    await tokenRecord.update({ revoked: true });
    throw new Error('Refresh token expired');
  }

  return { tokenRecord, payload };
}

async function revokeRefreshToken(token) {
  const hashedToken = crypto.createHash('sha256').update(token).digest('hex');
  const tokenRecord = await RefreshToken.findOne({ where: { token: hashedToken } });

  if (tokenRecord) {
    await tokenRecord.update({ revoked: true });
  }

  return true;
}

async function revokeAllUserTokens(userId) {
  await RefreshToken.update({ revoked: true }, { where: { userId, revoked: false } });
  return true;
}

module.exports = {
  generateAccessToken,
  generateRefreshToken,
  verifyAccessToken,
  verifyRefreshToken,
  revokeRefreshToken,
  revokeAllUserTokens,
};
